applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -895,276 +467,342 @@ static void cmd_send(struct k_work *work)
+	bool send = false;
+
+	for (processed = 0; processed < len && send == false; processed++) {
+
+		/* Handle control characters */
+		switch (buf[processed]) {
+		case 0x08: /* Backspace. */
+			/* Fall through. */
+		case 0x7F: /* DEL character */
+			if (at_cmd_len > 0) {
+				at_cmd_len--;
+			}
+			continue;
+		}
 
-	/* Handle control characters */
-	switch (character) {
-	case 0x08: /* Backspace. */
-		/* Fall through. */
-	case 0x7F: /* DEL character */
-		if (at_cmd_len > 0) {
-			at_cmd_len--;
+		/* Handle termination characters, if outside quotes. */
+		if (!inside_quotes) {
+			switch (buf[processed]) {
+			case '\r':
+				if (IS_ENABLED(CONFIG_SLM_CR_TERMINATION)) {
+					send = true;
+				}
+				break;
+			case '\n':
+				if (IS_ENABLED(CONFIG_SLM_LF_TERMINATION)) {
+					send = true;
+				} else if (IS_ENABLED(CONFIG_SLM_CR_LF_TERMINATION)) {
+					if (at_cmd_len > 0 && prev_character == '\r') {
+						at_cmd_len--; /* trim the CR char */
+						send = true;
+					}
+				}
+				break;
+			}
 		}
-		return 0;
-	}
 
-	/* Handle termination characters, if outside quotes. */
-	if (!inside_quotes) {
-		switch (character) {
-		case '\r':
-#if defined(CONFIG_SLM_CR_TERMINATION)
-			goto send;
-#else
-			break;
-#endif
-		case '\n':
-#if defined(CONFIG_SLM_LF_TERMINATION)
-			goto send;
-#elif defined(CONFIG_SLM_CR_LF_TERMINATION)
-			if (at_cmd_len > 0 && at_buf[at_cmd_len - 1] == '\r') {
-				at_cmd_len--; /* trim the CR char */
-				goto send;
+		if (send == false) {
+			/* Write character to AT buffer, leave space for null */
+			if (at_cmd_len < sizeof(at_buf) - 1) {
+				at_buf[at_cmd_len] = buf[processed];
 			}
-#endif
-			break;
+			at_cmd_len++;
+
+			/* Handle special written character */
+			if (buf[processed] == '"') {
+				inside_quotes = !inside_quotes;
+			}
+
+			prev_character = buf[processed];
 		}
+
 	}
 
-	/* Write character to AT buffer */
-	at_buf[at_cmd_len] = character;
-	at_cmd_len++;
+	if (send) {
+		if (at_cmd_len > sizeof(at_buf) - 1) {
+			LOG_ERR("AT command buffer overflow, %d dropped", at_cmd_len);
+			rsp_send_error();
+		} else if (at_cmd_len > 0) {
+			at_buf[at_cmd_len] = '\0';
+			cmd_send(at_buf, at_cmd_len, sizeof(at_buf));
+		} else {
+			/* Ignore 0 size command. */
+		}
 
-	/* Detect AT command buffer overflow, leaving space for null */
-	if (at_cmd_len > sizeof(at_buf) - 1) {
-		LOG_ERR("Buffer overflow");
-		at_cmd_len--;
-		at_buf_overflow = true;
-		goto send;
+		inside_quotes = false;
+		at_cmd_len = 0;
 	}
 
-	/* Handle special written character */
-	if (character == '"') {
-		inside_quotes = !inside_quotes;
+	return processed;
+}
+
+/* Search for quit_str and exit datamode when one is found. */
+static size_t null_handler(const uint8_t *buf, const size_t len)
+{
+	static size_t dropped_count;
+	static uint8_t match_count;
+
+	size_t processed;
+	bool match = false;
+
+	if (dropped_count == 0) {
+		LOG_ERR("Data pipe broken. Dropping data until datamode is terminated.");
```

Comment:
Going here when the function is called first time?

Reply:
Yes, that is the intention.